### PR TITLE
Assign scopes as a separate step.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

**Comment by chriseth** [comment]
The logic here is not proper yet, maybe not ever Scopable also opens a scope...

**Comment by ekpyron** [comment]
What about ``virtual bool ASTNode::opensScope() const { return false; }`` explicitly overridden in source unit, try catch, block, for, function and contract (if that's the exhaustive list, didn't check)?

**Comment by chriseth** [comment]
It turns out that the AST nodes that open scopes and those that store their scope are two completely different things.

Most notably, FunctionTypeName opens a scope but itself does not store the scope. This makes sense is a certain way but also leads to the fact that if you start with a variable declaration in a FunctionTypeName and follow the `scope()` pointers, you will not end up at the source unit.

File: libsolidity/analysis/NameAndTypeResolver.cpp [hunk 1]
```diff
@@ -633,7 +632,6 @@ void DeclarationRegistrationHelper::endVisit(FunctionTypeName&)
 
 bool DeclarationRegistrationHelper::visit(Block& _block)
```
Comment (chriseth) [inline libsolidity/analysis/NameAndTypeResolver.cpp]: These can be replaced by a generic visitor now.

File: libsolidity/ast/AST.h [hunk 2]
```diff
@@ -152,10 +152,19 @@ std::vector<T const*> ASTNode::filteredNodes(std::vector<ASTPointer<ASTNode>> co
 	return ret;
 }
 
+/**
+ * Abstract morker class that specifies that this AST node opens a scope.
```
Comment (axic) [inline libsolidity/ast/AST.h]: What is "morker"?

File: libsolidity/ast/AST.h [hunk 3]
```diff
@@ -1521,7 +1530,7 @@ class WhileStatement: public BreakableStatement
 /**
  * For loop statement
  */
-class ForStatement: public BreakableStatement, public Scopable
+class ForStatement: public BreakableStatement, public Scopable, public ScopeOpener
```
Comment (axic) [inline libsolidity/ast/AST.h:1538]: Hm, the while statement does not create its own scope? Or the scope of the while statement is represented by a block, and here we account for the variables defined in the loop init?

Comment (axic) [inline libsolidity/ast/AST.h:1538]: And what is the difference between Scopable and ScopeOpener?

Comment (chriseth) [reply libsolidity/ast/AST.h:1538]: No, its block does. For the for loop, we need a new scope because of the init part which is not a block.

Comment (chriseth) [reply libsolidity/ast/AST.h:1538]: Scopable stores the scope it is defined in and scope opener defines a new scope.

File: libsolidity/analysis/Scoper.h [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <libsolidity/ast/ASTForward.h>
+#include <libsolidity/ast/ASTVisitor.h>
+
+namespace solidity::langutil
+{
+class ErrorReporter;
```
Comment (hrkrshnn) [inline libsolidity/analysis/Scoper.h]: This doesn't seem to be used. Do we need it?

**Review by hrkrshnn — COMMENTED** [review]
Is it possible to merge both `VariableScope` and `ScopeOpener` into a single class?

**Comment by chriseth** [comment]
I don't think they can be merged, because not every ScopeOpener is a VariableScope.

File: libsolidity/analysis/Scoper.h (hunk 4 above)
Comment (chriseth) [reply libsolidity/analysis/Scoper.h]: Removed.

File: libsolidity/analysis/NameAndTypeResolver.cpp [hunk 5]
```diff
@@ -673,32 +566,44 @@ void DeclarationRegistrationHelper::endVisit(VariableDeclarationStatement& _vari
...
+	if (auto const* scopable = dynamic_cast<Scopable const*>(&_node))
+		solAssert(scopable->annotation().scope == m_currentScope, "");
+
+	if (auto* declaration = dynamic_cast<Declaration*>(&_node))
+		registerDeclaration(*declaration);
+	if (dynamic_cast<ScopeOpener const*>(&_node))
+		enterNewSubScope(_node);
+
+	if (dynamic_cast<ScopeOpener const*>(&_node))
+		m_currentScope = &_node;
```
Comment (hrkrshnn) [inline libsolidity/analysis/NameAndTypeResolver.cpp]: Isn't the scope already assigned when we call `enterNewScope(_node)`? So this assignment would be redundant, wouldn't it?

File: libsolidity/analysis/NameAndTypeResolver.cpp [hunk 6]
```diff
@@ -673,32 +566,44 @@ void DeclarationRegistrationHelper::endVisit(VariableDeclarationStatement& _vari
...
 {
-	map<ASTNode const*, shared_ptr<DeclarationContainer>>::iterator iter;
-	bool newlyAdded;
 	shared_ptr<DeclarationContainer> container{make_shared<DeclarationContainer>(m_currentScope, m_scopes[m_currentScope].get())};
-	tie(iter, newlyAdded) = m_scopes.emplace(&_subScope, move(container));
-	solAssert(newlyAdded, "Unable to add new scope.");
+	bool newlyAdded = m_scopes.emplace(&_subScope, move(container)).second;
+	// Source units are the only AST nodes for which containers can be created from multiple places
+	// due to imports.
+	solAssert(newlyAdded || dynamic_cast<SourceUnit const*>(&_subScope), "Unable to add new scope.");
```
Comment (hrkrshnn) [inline libsolidity/analysis/NameAndTypeResolver.cpp:604]: What would be an example where `newlyAdded` would be false?

File: libsolidity/analysis/NameAndTypeResolver.cpp (hunk 5 above)
Comment (chriseth) [reply libsolidity/analysis/NameAndTypeResolver.cpp]: That is true, thanks!

File: libsolidity/analysis/NameAndTypeResolver.cpp (hunk 6 above)
Comment (chriseth) [reply libsolidity/analysis/NameAndTypeResolver.cpp:604]: In an import directive, the scope of the imported file might already be created: https://github.com/ethereum/solidity/pull/9586/files#diff-653a1e23aa75715b13289b03e2f20947R536

**Review by hrkrshnn — APPROVED** [review]
LGTM. But one more review would be nice since I'm not so confident about this.

File: libsolidity/analysis/Scoper.cpp [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+#include <libsolidity/ast/AST.h>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::frontend;
+
+void Scoper::assignScopes(ASTNode const& _astRoot)
+{
+	Scoper scoper;
+	_astRoot.accept(scoper);
```
Comment (leonardoalt) [inline libsolidity/analysis/Scoper.cpp:30]: Why a new local variable instead of `this`?

Comment (leonardoalt) [reply libsolidity/analysis/Scoper.cpp:30]: Ah, because of the members?

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**